imodqgis/idf/conversion.py

```diff
@@ -0,0 +1,302 @@
+from pathlib import Path
+from typing import Any, Dict, Tuple, Union
+
+import struct
+import numpy as np
+
+from osgeo import gdal
+
+
+class NewGeoTiff:
+    """
+    Implements enter and exit to make sure the GDAL dataset is closed.
+    """
+
+    def __init__(self, path: str, nrow: int, ncol: int, dtype: type):
+        self.__raster__ = None
+        self.path = path
+        self.nrow = nrow
+        self.ncol = ncol
+        if dtype == "float32":
+            self.gdal_dtype = gdal.GDT_Float32
+        elif dtype == "float64":
+            self.gdal_dtype = gdal.GDT_Float64
+        else:
+            raise TypeError(f"IDF dtype should be float32 or float64. Got {dtype}")
+
+    def __enter__(self):
+        driver = gdal.GetDriverByName("GTiff")
+        self.__raster__ = driver.Create(
+            self.path, self.ncol, self.nrow, 1, self.gdal_dtype
+        )
+        return self
+
+    def __exit__(self, *_):
+        self.__raster__ = None
+
+    def set_transform(self, xmin: float, dx: float, ymax: float, dy: float) -> None:
+        self.__raster__.SetGeoTransform((xmin, dx, 0.0, ymax, 0.0, dy))
+
+    def set_crs(self, crs_wkt: str) -> None:
+        self.__raster__.SetProjection(crs_wkt)
+
+    def write_array(self, values: np.ndarray, nodata: float) -> None:
+        band = self.__raster__.GetRasterBand(1)
+        band.SetNoDataValue(nodata)
+        band.WriteArray(values)
+        band.FlushCache()
+
+
+class ReadOnlyRaster:
+    """
+    Implements enter and exit to make sure the GDAL dataset is closed.
+    """
+
+    def __init__(self, path: str):
+        self.__raster__ = None
+        self.path = path
+
+    def __enter__(self):
+        self.__raster__ = gdal.Open(self.path)
+        return self
+
+    def __exit__(self, *_):
+        self.__raster__ = None
+
+    def get_transform(self) -> Tuple[float]:
+        return self.__raster__.GetGeoTransform()
+
+    def read_array(self) -> Tuple[np.ndarray, Union[int, float, None]]:
+        band = self.__raster__.GetRasterBand(1)
+        values = band.ReadAsArray()
+        nodata = band.GetNoDataValue()
+        return values, nodata
+
+    @property
+    def nband(self) -> int:
+        return self.__raster__.RasterCount
+
+
+def read_idf(path: str) -> Tuple[Dict[str, Any], np.ndarray]:
+    """Read the IDF header information into a dictionary"""
+    attrs = {}
+    with open(path, "rb") as f:
+        reclen_id = struct.unpack("i", f.read(4))[0]  # Lahey RecordLength Ident.
+        if reclen_id == 1271:
```

Comment:
You could store this stuff in a constant, e.g. RECLEN_SINGLE

Or maybe a dataclass with e.g. and instance ``SINGLE``

```python

from dataclasses import dataclass


@dataclass
class Precision:
    reclen: int
    floatsize: int
    intsize: int
    floatformat: str
    intformat: str
    dtype: str
    doubleprecision: bool

SINGLE = Precision(
    reclen=1271,
    floatsize=4,
    intsize=4,
    floatformat="f",
    intformat="i",
    dtype="float32",
    doubleprecision=False,
)

```

Reply:
I think this is more hassle than it's worth, we just need the variables in the function body.